### DISCUSSION TITLE
Is there a way to INSERT multiple rows in oracle ?

### DISCUSSION THREAD

**Discussion by AbdealiLoKo** [sqlalchemy/sqlalchemy#10143] [Usage Questions] [answered]
I was using to insert rows into my table 1 row at a time:
```python
data = [{'col1': 'val1', 'col2': 'val2'}, ...]
for row in data:
    conn.execute(table.insert().values(row))
```

As the number of rows increased - understandably ... this became very slow.
So, I was trying to change it to:

```python
conn.execute(table.insert().values(data))
```
But this gave me an error in oracle - `CompileError: The 'oracle' dialect with current database version settings does not support in-place multirow inserts.`

But from what I read about oracle, this is something that is doable with `INSERT ALL` or some alternative syntaxes/workarounds
Reference: https://database.guide/4-ways-to-insert-multiple-rows-in-oracle/

Is there some way to insert multiple rows in 1 query in oracle that sqlalchemy suggests using ?

**Comment by CaselIT** [comment, marked as answer]
Hi,

Have you tried using `connection.execute(table.insert(), data)`?

**Reply by AbdealiLoKo** [reply]
5x faster than my earlier code - awesome ! 

On reading documentation ... I see this uses the `executemany` option in the underlying cursor.
Did not know that it existed :) Out of curiosity ...
On other DBs:
 - Is there a way for me to check if `executemany` is available ?
 - Is there a benefit of using executemany vs multirow-inserts ?

**Reply by CaselIT** [reply]
- Execute many should be supported by all dbapi
- depending on the driver, values may be a lot faster than execute many. A classic example is psycopg2

**Comment by CarlaFernandez** [comment]
Hello @AbdealiLoKo , would you mind sharing the code to create `conn`? I asumed it was the sqlAlchemy engine, but I'm getting an exception: `AttributeError: 'Engine' object has no attribute 'execute'`. Thanks!

**Reply by CarlaFernandez** [reply]
Also, please share what `table` is. I assumed it was the SQLAlchemy model class but it's not.

**Reply by CaselIT** [reply]
I would suggest you read (or at least skim) the sqlalchemy tutorial, in particular
- this part should explain what a connection is https://docs.sqlalchemy.org/en/20/tutorial/dbapi_transactions.html
- this instead talks about tables https://docs.sqlalchemy.org/en/20/tutorial/metadata.html

**Reply by AbdealiLoKo** [reply]
Something along the lines of:


```py
engine = create_engine(...)
with engine.connect() as conn:
    metadata = MetaData()
    metadata.reflect(engine)
    for table in metadata.sorted_tables:
        conn.execute(table.insert().values([...]))

```
